renew_apple_cert/create_csr.go

```diff
@@ -0,0 +1,134 @@
+package main
+
+import (
+	"crypto/rand"
+	"crypto/rsa"
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"encoding/asn1"
+	"encoding/pem"
+	"io/ioutil"
+	"log"
+	"os"
+
+	"github.com/joho/godotenv"
+)
+
+type inputCSR struct {
+	app            string
+	applePushTopic string
+	country        string
+	province       string
+	locality       string
+	organization   string
+	email          string
+}
+
+func newInputCSR(app string, applePushTopic string, country string, province string, locality string, organization string, email string) (i *inputCSR) {
+	i = &inputCSR{
+		app:            app,
+		applePushTopic: applePushTopic,
+		country:        country,
+		province:       province,
+		locality:       locality,
+		organization:   organization,
+		email:          email,
+	}
+	return i
+}
+
+func main() {
+	err := godotenv.Load()
+	if err != nil {
+		log.Fatal("Error loading .env file")
+	}
+
+	dirCsr := os.Getenv("DIR_CSR")
+	dirDownloaded := os.Getenv("DIR_DOWNLOADED")
+	dirs := []string{
+		dirCsr,
+		dirDownloaded,
+	}
+	for _, dir := range dirs {
+		_, err = os.Stat(dir)
+		if os.IsNotExist(err) {
+			err = os.MkdirAll(dir, 0644)
+			if err != nil {
+				panic(err)
+			}
+		}
+	}
+
+	i := newInputCSR(
+		os.Getenv("APP"),
+		os.Getenv("APPLE_PUSH_TOPIC"),
+		os.Getenv("COUNTRY"),
+		os.Getenv("PROVINCE"),
+		os.Getenv("LOCALITY"),
+		os.Getenv("ORGANIZATION"),
+		os.Getenv("EMAIL"),
+	)
+
+	key, err := createAndWritePrivateKey(i.app, dirCsr)
+	if err != nil {
+		panic(err)
+	}
+
+	err = createAndWriteCSR(i, key, dirCsr)
+	if err != nil {
+		panic(err)
+	}
+}
+
+func createAndWritePrivateKey(app string, dirCsr string) (key *rsa.PrivateKey, err error) {
+	key, err = rsa.GenerateKey(rand.Reader, 2048)
+	if err != nil {
+		return nil, err
+	}
+	marshaledKey := x509.MarshalPKCS1PrivateKey(key)
+	pemPrivateKey := pem.EncodeToMemory(
+		&pem.Block{
+			Type:  "RSA PRIVATE KEY",
+			Bytes: marshaledKey,
+		},
+	)
+	err = ioutil.WriteFile(dirCsr+"/"+app+".key", pemPrivateKey, 0644)
```

Comment:
In case this is ever run in a multi-user environment: surely the private key and everything related to it should be `0600` max? That applies to parent directories, conversion output etc.

Reply:
Agreed.